servlet/src/main/scala/org/http4s/servlet/ServletIo.scala

```diff
@@ -19,51 +19,72 @@ package servlet
 
 import cats.effect._
 import cats.effect.std.Dispatcher
-import cats.effect.std.Queue
 import cats.syntax.all._
 import fs2._
+import org.http4s.internal.bug
 import org.log4s.getLogger
 
-import java.util.Arrays
+import java.util.concurrent.CancellationException
+import java.util.concurrent.atomic.AtomicReference
 import javax.servlet.ReadListener
 import javax.servlet.WriteListener
 import javax.servlet.http.HttpServletRequest
 import javax.servlet.http.HttpServletResponse
+import scala.annotation.tailrec
 
 /** Determines the mode of I/O used for reading request bodies and writing response bodies.
   */
-sealed trait ServletIo[F[_]] {
+sealed abstract class ServletIo[F[_]: Async] {
+  protected[servlet] val F: Async[F] = Async[F]
+
+  protected[servlet] def reader(servletRequest: HttpServletRequest): EntityBody[F]
+
+  /** An alias for [[reader]].  In the future, this will be optimized with
+    * the dispatcher.
+    *
+    * @param dispatcher currently ignored
+    */
   def requestBody(
       servletRequest: HttpServletRequest,
       dispatcher: Dispatcher[F],
-  ): Stream[F, Byte]
+  ): Stream[F, Byte] = {
+    val _ = dispatcher // unused
+    reader(servletRequest)
+  }
 
+  /** May install a listener on the servlet response. */
+  protected[servlet] def initWriter(servletResponse: HttpServletResponse): BodyWriter[F]
+
+  /** An alias for [[initWriter]].  In the future, this will be
+    * optimized with the dispatcher.
+    *
+    * @param dispatcher currently ignored
+    */
   def bodyWriter(servletResponse: HttpServletResponse, dispatcher: Dispatcher[F])(
       response: Response[F]
-  ): F[Unit]
+  ): F[Unit] = {
+    val _ = dispatcher
+    initWriter(servletResponse)(response)
+  }
 }
 
 /** Use standard blocking reads and writes.
   *
   * This is more CPU efficient per request than [[NonBlockingServletIo]], but is likely to
   * require a larger request thread pool for the same load.
   */
-final case class BlockingServletIo[F[_]](chunkSize: Int)(implicit F: Sync[F]) extends ServletIo[F] {
-  override def requestBody(
-      servletRequest: HttpServletRequest,
-      dispatcher: Dispatcher[F],
-  ): EntityBody[F] =
+final case class BlockingServletIo[F[_]: Async](chunkSize: Int) extends ServletIo[F] {
```

Comment:
This can probably keep the `Sync` constraint.

Reply:
Done. Upcoming PR